### PR TITLE
Ensure `_prepare_state` support jax

### PR DESCRIPTION
**Pull request by Ericgig**
**Description**
The check that the initial state was already normalised for normalising output would branch over values, which breaks `jax.jit` support... Moved the option check first.

Also changed the `isoper` check to filter out rectangular operator.
Rectangular state in solver are tolerated, but we don't have tests for it. I did not add any here, but we should 

**Related issues or PRs**
qutip/qutip-jax#49 qutip/qutip-tutorials#100

### CODE REVIEW

**Review by pmenczel — APPROVED** [review]
Looks good to me, and I think this should fix the issues in qutip-qoc
